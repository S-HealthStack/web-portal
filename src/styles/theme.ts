import { css, DefaultTheme } from 'styled-components';
import { px } from './utils';

const colors = {
  // gray gradations
  black: '#000000',
  black40: 'rgba(0, 0, 0, 0.4)',
  black08: 'rgba(0, 0, 0, 0.08)',

  // light mode
  surface: '#FFFFFF',
  background: '#F7F8FA',
  onPrimary: '#FFFFFF',
  backgroundOnPrimary: '#FFFFFF',
  backgroundSurface: '#FFFFFF',
  backgroundOnSurface: '#2D2D2D',
  primary: '#4475E3',
  primaryLight: '#ECF1FC',
  primaryHovered: '#2F65E0',
  primary05: '#F6F8FE',
  primary10: '#ECF1FC',
  primary30: '#C7D6F7',
  primaryDisabled: '#B3C6F1',
  primaryLightPressed: '#E3EAFB',
  primaryLightFocused: '#C7D6F7',
  textPrimary: '#474747',
  primaryWhite: '#FFFFFF',
  disabled: '#DBDBDB',
  disabled20: 'rgba(219, 219, 219, 0.2)',
  onDisabled: '#A1A1A1',

  textPrimaryBlue: '#4475E3',
  textPrimaryDark: '#474747',
  textSecondaryGray: '#808080',
  textDisabled: '#838383',
  onSurface: '#2D2D2D',
  secondarySkyblue: '#00B0D7', // replace
  backgoundLight: '#F0F0F0',
  primaryBlueHovered: '#2F65E0',
  primaryBluePressed: '#2057D5',

  statusSuccess: '#2DC008',
  statusSuccess10: '#DCF2EA',
  statusSuccessText: '#00825D',

  statusError: '#FF3333',
  statusError10: '#FFEBEB',
  statusErrorText: '#D14343',

  statusWarning: '#F5BF00',
  statusWarning10: '#FFF7D0',
  statusWarningText: '#A36400',

  secondaryViolet: '#944ED7',
  secondaryVioletHovered: '#6A3799',
  secondaryViolet10: '#F4EDFB',

  secondaryTangerine: '#F39C18',
  secondaryTangerineHovered: '#CC8314',
  secondaryTangerine10: '#FEF5E8',

  secondarySkyBlue: '#00B0D7',
  secondarySkyBlueHovered: '#0092B3',
  secondarySkyBlue10: '#E5F7FB',

  secondaryGreen: '#1AD598',
  secondaryGreenHovered: '#16B37F',
  secondaryGreen10: '#E8FBF5',

  secondaryRed: '#FA5F4F',
  secondaryRedHovered: '#CC4E40',
  secondaryRed10: '#FFEFED',
};

export type SpecColorType = keyof typeof colors;

const weightToCssValue = {
  regular: 400,
  upperregular: 450,
  medium: 500,
  semibold: 600,
  bold: 700,
};

const fontSpec = (
  size: number,
  lineHeight: number | string,
  weight: keyof typeof weightToCssValue,
  fontFamily: string
) => css`
  font-family: ${fontFamily}, sans-serif;
  font-size: ${px(size)};
  line-height: ${typeof lineHeight === 'string' ? lineHeight : px(lineHeight)};
  font-weight: ${weightToCssValue[weight]};
  font-style: normal;
`;

const typography = {
  // Inter typeface
  labelRegular: fontSpec(10, 15, 'regular', 'Inter'),
  labelSemibold: fontSpec(10, 15, 'semibold', 'Inter'),

  bodyXSmallRegular: fontSpec(12, 18, 'regular', 'Inter'),
  bodyXSmallSemibold: fontSpec(12, 18, 'semibold', 'Inter'),
  bodySmallRegular: fontSpec(14, 21, 'regular', 'Inter'),
  bodySmallSemibold: fontSpec(14, 21, 'semibold', 'Inter'),
  bodyMediumRegular: fontSpec(16, 24, 'regular', 'Inter'),
  bodyMediumSemibold: fontSpec(16, 24, 'semibold', 'Inter'),
  bodyLargeRegular: fontSpec(18, 26, 'regular', 'Inter'),
  bodyLargeSemibold: fontSpec(18, 26, 'semibold', 'Inter'),

  headingXSmall: fontSpec(16, 21, 'semibold', 'Inter'),
  headingSmall: fontSpec(18, 23, 'semibold', 'Inter'),
  headingXMedium: fontSpec(18, 27, 'semibold', 'Inter'),
  headingXMediumRegular: fontSpec(18, 27, 'regular', 'Inter'),
  headingMedium: fontSpec(24, 32.2, 'semibold', 'Inter'),
  headingLargeSemibold: fontSpec(32, 42, 'semibold', 'Inter'),
  headingXLargeSemibold: fontSpec(48, 58, 'semibold', 'Inter'),
  headingXXLargeSemibold: fontSpec(60, 72, 'semibold', 'Inter'),

  // Fira Code typeface
  query14: fontSpec(14, '130%', 'upperregular', 'Fira Code'),
  query12: fontSpec(12, '150%', 'upperregular', 'Fira Code'),
};

const boxShadow = {
  card: `0 0 ${px(2)} ${px(0)} rgba(0,0,0,0.15)`,
  avatar: `0 ${px(8)} ${px(10)} ${px(2)} 0 rgba(0,0,0,0.12)`,
  studyAvatar: `0 ${px(15)} ${px(20)} rgba(180,180,180,0.3)`,
  previewScreen: `0 ${px(2)} ${px(4)} rgba(71,71,71,0.25)`,
};

const animation = {
  defaultTiming: 'cubic-bezier(0.45, 0.05, 0, 1)',
  rippleTiming: 'cubic-bezier(0.4, 0, 0.2, 1)',
};

declare module 'styled-components' {
  export interface DefaultTheme {
    colors: typeof colors;
    typography: typeof typography;
    boxShadow: typeof boxShadow;
    animation: typeof animation;
  }
}

const theme: DefaultTheme = {
  colors,
  typography,
  boxShadow,
  animation,
};

export default theme;
