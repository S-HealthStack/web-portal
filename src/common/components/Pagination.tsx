import React, { FC, useCallback, useMemo } from 'react';

import styled from 'styled-components';

import Button from 'src/common/components/Button';
import Dropdown, { MenuItem, ValueItem } from 'src/common/components/Dropdown';
import PageFirstIcon from 'src/assets/icons/page_first.svg';
import PageLastIcon from 'src/assets/icons/page_last.svg';
import ChevronLeftIcon from 'src/assets/icons/chevron_left_pagination.svg';
import ChevronRightIcon from 'src/assets/icons/chevron_right_pagination.svg';
import { colors, px, typography } from 'src/styles';

export const PAGINATION_LIMITS = [10, 25, 50, 75];

export interface PaginationProps {
  totalCount: number;
  pageSize: number;
  onPageChange: (offset: number, size: number) => void;
  disabled?: boolean;
  offset: number;
}

const Container = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  ${typography.bodySmallRegular};
  line-height: ${px(28)};
`;

const Info = styled.div`
  ${typography.bodySmallRegular};
  color: ${colors.primary};
  line-height: ${px(28)};
`;

const Controls = styled.div`
  align-items: center;
  justify-content: flex-end;
  display: flex;
  margin-right: ${px(-2)};
`;

const ArrowButton = styled(Button).attrs({
  fill: 'text',
  rate: 'icon',
})`
  margin-left: ${px(16)};
  &:disabled {
    > div {
      background-color: transparent;
      svg {
        fill: ${colors.disabled};
      }
    }
  }
`;

const DropDownContainer = styled.div`
  display: flex;
  align-items: center;
  margin-right: ${px(41)};
`;

const DropDownLabel = styled.span`
  ${typography.bodySmallRegular};
  color: ${colors.primary};
  margin-right: ${px(8)};
`;

const DropDownControl = styled(Dropdown)`
  width: ${px(40)};
  height: ${px(25)};
  border: none;
  border-radius: ${px(4)};
  z-index: 10;
  color: ${colors.textSecondaryGray};

  & > ${ValueItem} {
    position: relative;
    z-index: 100;
    display: flex;
    align-items: center;
    justify-content: center;
    padding-left: ${px(6)};
    padding-right: ${px(6)};
    border-color: ${colors.disabled};
    &:hover {
      cursor: pointer;
    }
    > svg {
      display: none;
    }
  }
` as typeof Dropdown;

const SmallMenuItem = styled(MenuItem)`
  background-color: ${({ selected }) => (selected ? colors.primary10 : colors.backgroundSurface)};
  width: ${px(38)};
  padding: 0;
  overflow-x: hidden;

  div {
    width: ${px(38)};
    text-align: center;
    > svg {
      display: none;
    }
  }
`;

const Pagination: FC<PaginationProps> = ({
  totalCount,
  pageSize,
  onPageChange,
  disabled,
  offset,
  ...props
}) => {
  const isPrevDisabled = offset === 0;
  const isNextDisabled = offset + pageSize >= totalCount;

  const handlePageClick = useCallback(
    (newOffset: number) => !disabled && onPageChange(newOffset, pageSize),
    [pageSize, onPageChange, disabled]
  );

  const handleSizeChanged = useCallback(
    (size: number) => {
      if (!disabled) {
        onPageChange(offset, size);
      }
    },
    [onPageChange, disabled, offset]
  );

  const sizes = useMemo(() => PAGINATION_LIMITS.map((i) => ({ label: i.toString(), key: i })), []);

  const goToFirst = useCallback(() => handlePageClick(0), [handlePageClick]);
  const goToLast = useCallback(
    () => handlePageClick(totalCount - pageSize),
    [handlePageClick, totalCount, pageSize]
  );

  const goToPrev = useCallback(
    () => handlePageClick(Math.max(offset - pageSize, 0)),
    [handlePageClick, offset, pageSize]
  );
  const goToNext = useCallback(
    () => handlePageClick(Math.min(offset + pageSize, totalCount)),
    [handlePageClick, offset, pageSize, totalCount]
  );

  return (
    <Container {...props} data-testid="pagination">
      <Info data-testid="info" {...props}>{`${offset + 1}-${Math.min(
        offset + pageSize + 1,
        totalCount
      )} of ${totalCount}`}</Info>
      <Controls {...props}>
        <DropDownContainer>
          <DropDownLabel>Rows per page</DropDownLabel>
          <DropDownControl
            items={sizes}
            activeKey={pageSize}
            onChange={handleSizeChanged}
            menuItemComponent={SmallMenuItem}
            backgroundType="light"
            menuItemHeight={25}
            maxVisibleMenuItems={5}
          />
        </DropDownContainer>
        <ArrowButton
          disabled={isPrevDisabled}
          onClick={goToFirst}
          data-testid="go-to-first"
          aria-label="Go to First"
        >
          <PageFirstIcon />
        </ArrowButton>
        <ArrowButton
          disabled={isPrevDisabled}
          onClick={goToPrev}
          data-testid="go-to-previous"
          aria-label="Go to Previous"
        >
          <ChevronLeftIcon />
        </ArrowButton>
        <ArrowButton
          disabled={isNextDisabled}
          onClick={goToNext}
          data-testid="go-to-next"
          aria-label="Go to Next"
        >
          <ChevronRightIcon />
        </ArrowButton>
        <ArrowButton
          disabled={isNextDisabled}
          onClick={goToLast}
          data-testid="go-to-last"
          aria-label="Go to Last"
        >
          <PageLastIcon />
        </ArrowButton>
      </Controls>
    </Container>
  );
};

export default Pagination;
