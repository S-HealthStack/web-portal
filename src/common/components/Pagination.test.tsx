import React from 'react';
import '@testing-library/jest-dom';
import 'jest-styled-components';
import { render } from '@testing-library/react';
import theme from 'src/styles/theme';
import userEvent from '@testing-library/user-event';
import { ThemeProvider } from 'styled-components/';
import Pagination from './Pagination';

describe('Pagination', () => {
  const TOTAL = 1000;
  const PAGE_SIZE = 10;
  const CURRENT_OFFSET = 0;

  it('test pagination render', async () => {
    const onPageChange = jest.fn();

    const { baseElement, queryByTestId } = render(
      <ThemeProvider theme={theme}>
        <Pagination
          totalCount={TOTAL}
          pageSize={PAGE_SIZE}
          offset={CURRENT_OFFSET}
          onPageChange={onPageChange}
        />
      </ThemeProvider>
    );

    expect(baseElement).toMatchSnapshot();

    const pagination = queryByTestId('pagination') as Element;
    const prevButton = queryByTestId('go-to-previous') as Element;
    const nextButton = queryByTestId('go-to-next') as Element;
    const firstButton = queryByTestId('go-to-first') as Element;
    const lastButton = queryByTestId('go-to-last') as Element;
    const info = queryByTestId('info') as Element;

    expect(pagination).toBeInTheDocument();
    expect(prevButton).toBeInTheDocument();
    expect(nextButton).toBeInTheDocument();
    expect(firstButton).toBeInTheDocument();
    expect(lastButton).toBeInTheDocument();
    expect(info).toBeInTheDocument();

    expect(prevButton).toBeDisabled();
    expect(firstButton).toBeDisabled();
    expect(nextButton).not.toBeDisabled();
    expect(lastButton).not.toBeDisabled();
    expect(info).toHaveTextContent(`${CURRENT_OFFSET + 1}-${PAGE_SIZE + 1} of ${TOTAL}`);

    await userEvent.click(nextButton);

    expect(onPageChange).toHaveBeenCalledTimes(1);
  });

  it('[NEGATIVE] should render with wrong props', async () => {
    const onPageChange = jest.fn();

    const { baseElement, queryByTestId } = render(
      <ThemeProvider theme={theme}>
        <Pagination totalCount={-1} pageSize={-1} offset={-1} onPageChange={onPageChange} />
      </ThemeProvider>
    );

    expect(baseElement).toMatchSnapshot();

    const pagination = queryByTestId('pagination') as Element;
    const prevButton = queryByTestId('go-to-previous') as Element;
    const nextButton = queryByTestId('go-to-next') as Element;
    const firstButton = queryByTestId('go-to-first') as Element;
    const lastButton = queryByTestId('go-to-last') as Element;

    expect(pagination).toBeInTheDocument();
    expect(prevButton).toBeInTheDocument();
    expect(nextButton).toBeInTheDocument();
    expect(firstButton).toBeInTheDocument();
    expect(lastButton).toBeInTheDocument();
  });
});
