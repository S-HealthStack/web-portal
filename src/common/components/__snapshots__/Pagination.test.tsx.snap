// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination test pagination render 1`] = `
.c12 {
  position: relative;
  display: -webkit-box !important;
  display: -webkit-flex !important;
  display: -ms-flexbox !important;
  display: flex !important;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border-radius: 0.25rem;
  width: 100%;
  white-space: nowrap;
  padding: 0;
  overflow: hidden;
  -webkit-transition-duration: 0;
  transition-duration: 0;
  -webkit-transition-property: background-color;
  transition-property: background-color;
  height: 1.5rem;
  width: 1.5rem;
  border: none;
  background-color: transparent;
}

.c12 > div {
  color: #4475E3;
}

.c12 > div svg {
  fill: #4475E3;
}

.c12:hover:enabled > div {
  color: #2F65E0;
}

.c12:hover:enabled > div svg {
  fill: #2F65E0;
}

.c12:hover:enabled:active > div {
  color: #2057D5;
}

.c12:hover:enabled:active > div svg {
  fill: #2057D5;
}

.c12:disabled > div {
  color: #B3C6F1;
}

.c12:disabled > div svg {
  fill: #B3C6F1;
}

.c12:focus-visible {
  outline: 0.125rem solid #C7D6F7;
}

.c12:focus-visible > div {
  color: #4475E3;
}

.c12:focus-visible > div svg {
  fill: #4475E3;
}

.c12:hover:enabled {
  background-color: #ECF1FC;
}

.c12:hover:enabled:active > div {
  color: #2057D5;
}

.c12:hover:enabled:active > div svg {
  fill: #2057D5;
}

.c12:focus-visible:enabled {
  background-color: #C7D6F7;
  outline: none;
}

.c15 {
  position: relative;
  display: -webkit-box !important;
  display: -webkit-flex !important;
  display: -ms-flexbox !important;
  display: flex !important;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border-radius: 0.25rem;
  width: 100%;
  white-space: nowrap;
  padding: 0;
  overflow: hidden;
  -webkit-transition-duration: 0;
  transition-duration: 0;
  -webkit-transition-property: background-color;
  transition-property: background-color;
  height: 1.5rem;
  width: 1.5rem;
  border: none;
  background-color: transparent;
}

.c15:hover {
  cursor: pointer;
}

.c15 > div {
  color: #4475E3;
}

.c15 > div svg {
  fill: #4475E3;
}

.c15:hover:enabled > div {
  color: #2F65E0;
}

.c15:hover:enabled > div svg {
  fill: #2F65E0;
}

.c15:hover:enabled:active > div {
  color: #2057D5;
}

.c15:hover:enabled:active > div svg {
  fill: #2057D5;
}

.c15:disabled > div {
  color: #B3C6F1;
}

.c15:disabled > div svg {
  fill: #B3C6F1;
}

.c15:focus-visible {
  outline: 0.125rem solid #C7D6F7;
}

.c15:focus-visible > div {
  color: #4475E3;
}

.c15:focus-visible > div svg {
  fill: #4475E3;
}

.c15:hover:enabled {
  background-color: #ECF1FC;
}

.c15:hover:enabled:active > div {
  color: #2057D5;
}

.c15:hover:enabled:active > div svg {
  fill: #2057D5;
}

.c15:focus-visible:enabled {
  background-color: #C7D6F7;
  outline: none;
}

.c14 {
  position: relative;
  z-index: 2;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transform: translate3d(0,0,0.0625rem);
  -ms-transform: translate3d(0,0,0.0625rem);
  transform: translate3d(0,0,0.0625rem);
}

.c5 {
  position: relative;
  width: 100%;
}

.c10 {
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 400;
  font-style: normal;
  color: #474747;
  grid-area: label;
}

.c11 {
  grid-area: right-icon;
}

.c8 {
  width: 100%;
  height: 1.5625rem;
  padding: 0 1rem;
  padding-right: 0.5rem;
  box-sizing: border-box;
  display: grid;
  grid-template-areas: ' label right-icon';
  grid-template-columns: 1fr 24px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 0.25rem;
  background-color: #FFFFFF;
  border: 0.0625rem solid transparent;
  border-top-width: 0;
  -webkit-transition: all 300ms cubic-bezier(0.45,0.05,0,1);
  transition: all 300ms cubic-bezier(0.45,0.05,0,1);
}

.c8:hover {
  cursor: pointer;
}

.c9 {
  border-top-width: 0.0625rem;
  border-color: !important;
  border-radius: 0.25rem;
}

.c9:hover {
  border-color: #2F65E0;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 400;
  font-style: normal;
  line-height: 1.75rem;
}

.c1 {
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 400;
  font-style: normal;
  color: #4475E3;
  line-height: 1.75rem;
}

.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: -0.125rem;
}

.c13 {
  margin-left: 1rem;
}

.c13:disabled > div svg {
  fill: #DBDBDB;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-right: 2.5625rem;
}

.c4 {
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 400;
  font-style: normal;
  color: #4475E3;
  margin-right: 0.5rem;
}

.c6 {
  width: 2.5rem;
  height: 1.5625rem;
  border: none;
  border-radius: 0.25rem;
  z-index: 10;
  color: #808080;
}

.c6 > .c7 {
  position: relative;
  z-index: 100;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding-left: 0.375rem;
  padding-right: 0.375rem;
  border-color: #DBDBDB;
}

.c6 > .c7:hover {
  cursor: pointer;
}

.c6 > .c7 > svg {
  display: none;
}

<body>
  <div>
    <div
      class="c0"
      data-testid="pagination"
    >
      <div
        class="c1"
        data-testid="info"
      >
        1-11 of 1000
      </div>
      <div
        class="c2"
      >
        <div
          class="c3"
        >
          <span
            class="c4"
          >
            Rows per page
          </span>
          <div
            class="c5 c6"
          >
            <div
              class="c7 c8 c9"
              data-testid="ValueItem"
            >
              <div
                class="c10"
              >
                10
              </div>
              <svg
                class="c11"
              />
            </div>
          </div>
        </div>
        <button
          class="c12 c13"
          data-testid="go-to-first"
          disabled=""
          fill="text"
        >
          <div
            class="c14"
            data-testid="content"
          >
            <svg />
          </div>
        </button>
        <button
          class="c12 c13"
          data-testid="go-to-previous"
          disabled=""
          fill="text"
        >
          <div
            class="c14"
            data-testid="content"
          >
            <svg />
          </div>
        </button>
        <button
          class="c15 c13"
          data-testid="go-to-next"
          fill="text"
        >
          <div
            class="c14"
            data-testid="content"
          >
            <svg />
          </div>
        </button>
        <button
          class="c15 c13"
          data-testid="go-to-last"
          fill="text"
        >
          <div
            class="c14"
            data-testid="content"
          >
            <svg />
          </div>
        </button>
      </div>
    </div>
  </div>
</body>
`;
